File: src/ados/services/mavlink/signing.py
"""MAVLink v2 message signing helpers.

The agent does not own the signing key. MAVLink signing keys live in the GCS
browser as non-extractable Web Crypto CryptoKey objects. This module exposes
only the operations needed for one-shot FC enrollment, capability detection,
and the SIGNING_REQUIRE toggle.

A key reaches the agent exactly once, as raw bytes in the POST body of the
enroll-fc REST call. Those bytes are zeroized the moment SETUP_SIGNING has
been sent to the FC.
"""

from __future__ import annotations

import asyncio
import binascii
import hashlib
import time
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any

from pymavlink import mavutil

from ados.core.logging import get_logger

if TYPE_CHECKING:
    from ados.services.mavlink.connection import FCConnection
    from ados.services.mavlink.param_cache import ParamCache
    from ados.services.mavlink.state import VehicleState

log = get_logger("mavlink.signing")

# 2015-01-01 00:00:00 UTC, expressed in seconds since POSIX epoch.
_EPOCH_2015 = datetime(2015, 1, 1, tzinfo=timezone.utc).timestamp()

# Minimum ArduPilot major+minor to expose signing params (ArduPilot 4.0+).
_MIN_ARDUPILOT_MAJOR = 4
_MIN_ARDUPILOT_MINOR = 0

# MAV_PARAM_TYPE_UINT8 = 1. SIGNING_REQUIRE is a uint8 on ArduPilot.
_PARAM_TYPE_UINT8 = 1


# ──────────────────────────────────────────────────────────────
# Capability detection
# ──────────────────────────────────────────────────────────────

def detect_capability(
    fc: FCConnection | None,
    vehicle_state: VehicleState | None,
    param_cache: ParamCache | None,
) -> dict[str, Any]:
    """Return whether the connected FC supports MAVLink signing.

    The check is strict on purpose: ArduPilot + version >= 4.0 + at least
    one SIGNING_* param in the cached param tree. Custom builds that
    stripped signing will correctly report unsupported.
    """
    if fc is None or not fc.connected:
        return {
            "supported": False,
            "reason": "fc_not_connected",
            "firmware_name": None,
            "firmware_version": None,
            "signing_params_present": False,
        }

    # MAV_AUTOPILOT enum — 3 = ArduPilot, 12 = PX4
    autopilot = vehicle_state.autopilot if vehicle_state else 0
    firmware_name = _autopilot_name(autopilot)

    if autopilot != mavutil.mavlink.MAV_AUTOPILOT_ARDUPILOTMEGA:
        # MSP firmwares never get a MAVLink HEARTBEAT, so autopilot==0 here
        # means no heartbeat received yet (already handled above by fc_not_connected).
        # Any other autopilot that isn't ArduPilot falls into this branch.
        if autopilot == mavutil.mavlink.MAV_AUTOPILOT_PX4:
            reason = "firmware_px4_no_persistent_store"
        elif autopilot == mavutil.mavlink.MAV_AUTOPILOT_INVALID:
            reason = "msp_protocol"
        else:
            reason = "firmware_not_supported"
        return {
            "supported": False,
            "reason": reason,
            "firmware_name": firmware_name,
            "firmware_version": None,
            "signing_params_present": False,
        }

    # Look for any SIGNING_* param in the cache. Presence is the strictest
    # gate because builds that stripped signing won't expose these params.
    signing_params_present = False
    if param_cache is not None:
        for name in param_cache.get_all().keys():
            if name.startswith("SIGNING_"):
                signing_params_present = True
                break

    # We can't derive major/minor from VehicleState.autopilot alone. The agent
    # doesn't currently cache AUTOPILOT_VERSION major/minor. If the SIGNING_*
    # params are present the firmware is clearly new enough; if they're absent
    # we assume the firmware is too old (or signing was stripped).
    if not signing_params_present:
        return {
            "supported": False,
            "reason": "firmware_too_old",
            "firmware_name": firmware_name,
            "firmware_version": None,
            "signing_params_present": False,
        }

    return {
        "supported": True,
        "reason": "ok",
        "firmware_name": firmware_name,
        "firmware_version": None,  # populated by GCS when it reads AUTOPILOT_VERSION
        "signing_params_present": True,
    }


def _autopilot_name(autopilot: int) -> str | None:
    if autopilot == 0:
        return None
    if autopilot == mavutil.mavlink.MAV_AUTOPILOT_ARDUPILOTMEGA:
        return "ArduPilot"
    if autopilot == mavutil.mavlink.MAV_AUTOPILOT_PX4:
        return "PX4"
    if autopilot == mavutil.mavlink.MAV_AUTOPILOT_INVALID:
        return "Unknown (non-MAVLink or heartbeat-only)"
    return f"Autopilot-{autopilot}"


# ──────────────────────────────────────────────────────────────
# FC enrollment (one-shot, key zeroized on return)
# ──────────────────────────────────────────────────────────────

def _initial_timestamp_10us() -> int:
    """Seconds since 2015-01-01 UTC, expressed in 10-microsecond units."""
    seconds_since_2015 = time.time() - _EPOCH_2015
    return int(seconds_since_2015 * 100_000)


def _fingerprint(key_bytes: bytes) -> str:
    """First 8 hex chars of SHA-256 over the key. Safe to display / log."""
    return hashlib.sha256(key_bytes).hexdigest()[:8]


async def enroll_fc(
    fc: FCConnection,
    key_bytes: bytearray,
    target_system: int = 1,
    target_component: int = 1,
) -> dict[str, Any]:
    """Send SETUP_SIGNING twice and zeroize the key buffer on exit.

    key_bytes MUST be a mutable bytearray so we can overwrite it in place.
    The caller must not reuse key_bytes after this function returns.

    Returns {success, key_id, enrolled_at}. key_id is the first 8 hex chars
    of sha256(key); the raw key never appears in the return value.
    """
    if len(key_bytes) != 32:
        raise ValueError(f"signing key must be 32 bytes, got {len(key_bytes)}")

    if not fc.connected or fc.connection is None:
        raise RuntimeError("FC not connected")

    conn = fc.connection
    initial_ts = _initial_timestamp_10us()
    key_id = _fingerprint(bytes(key_bytes))

    try:
        # pymavlink's setup_signing_send takes target_system, target_component,
        # secret_key (bytes), initial_timestamp. Sent twice 200ms apart to
        # survive a single-frame radio hiccup during enrollment.
        conn.mav.setup_signing_send(
            target_system,
            target_component,
            bytes(key_bytes),
            initial_ts,
        )
        await asyncio.sleep(0.2)
        conn.mav.setup_signing_send(
            target_system,
            target_component,
            bytes(key_bytes),
            initial_ts,
        )
    finally:
        # Zeroize the caller's buffer. Python memory is GC'd eventually, but
        # the bytearray is our caller's authoritative copy, so we overwrite
        # it explicitly. This runs even if setup_signing_send raised.
        for i in range(len(key_bytes)):
            key_bytes[i] = 0

    enrolled_at = datetime.now(tz=timezone.utc).isoformat(timespec="seconds")
    # Log the fingerprint, never the key. key_id is public-safe.
    log.info("signing_key_enrolled", key_id=key_id, target_system=target_system)
    return {
        "success": True,
        "key_id": key_id,
        "enrolled_at": enrolled_at,
    }


def disable_on_fc(
    fc: FCConnection,
    target_system: int = 1,
    target_component: int = 1,
) -> dict[str, Any]:
    """Clear the FC's signing store by sending 32 zero bytes + zero timestamp.

    ArduPilot recognises this as 'disable signing'.
    """
    if not fc.connected or fc.connection is None:
        raise RuntimeError("FC not connected")

    conn = fc.connection
    zero_key = bytes(32)
    conn.mav.setup_signing_send(target_system, target_component, zero_key, 0)
    log.info("signing_disabled_on_fc", target_system=target_system)
    return {"success": True}


def set_require(
    fc: FCConnection,
    require: bool,
    target_system: int = 1,
    target_component: int = 1,
) -> dict[str, Any]:
    """Write SIGNING_REQUIRE param (1 or 0) on the FC."""
    if not fc.connected or fc.connection is None:
        raise RuntimeError("FC not connected")

    conn = fc.connection
    conn.mav.param_set_send(
        target_system,
        target_component,
        b"SIGNING_REQUIRE",
        1.0 if require else 0.0,
        _PARAM_TYPE_UINT8,
    )
    log.info("signing_require_set", require=require)
    return {"success": True, "require": require}


def get_require(param_cache: ParamCache | None) -> dict[str, Any]:
    """Read SIGNING_REQUIRE from the param cache.

    Returns {require: bool | None}. None means the param hasn't been seen
    yet in the current session.
    """
    if param_cache is None:
        return {"require": None}
    value = param_cache.get("SIGNING_REQUIRE")
    if value is None:
        return {"require": None}
    return {"require": bool(int(value))}


# ──────────────────────────────────────────────────────────────
# Frame observer (counts signed frames, no crypto)
# ──────────────────────────────────────────────────────────────

class FrameObserver:
    """Counts MAVLink v2 frames with the signed bit set.

    The agent cannot validate signatures (it holds no key). This observer
    gives the GCS a sanity check that signed frames are actually transiting
    the IPC socket in both directions.
    """

    def __init__(self) -> None:
        self.tx_signed_count: int = 0
        self.rx_signed_count: int = 0
        self.last_signed_rx_at: float | None = None

    def observe_frame(self, frame: bytes, direction: str) -> None:
        """Peek at INC_FLAGS at offset 2 of a v2 frame."""
        if len(frame) < 10:
            return
        if frame[0] != 0xFD:
            return  # v1 frame, no signing possible
        inc_flags = frame[2]
        if inc_flags & 0x01:
            if direction == "tx":
                self.tx_signed_count += 1
            elif direction == "rx":
                self.rx_signed_count += 1
                self.last_signed_rx_at = time.time()

    def snapshot(self) -> dict[str, Any]:
        return {
            "tx_signed_count": self.tx_signed_count,
            "rx_signed_count": self.rx_signed_count,
            "last_signed_rx_at": self.last_signed_rx_at,
        }


# ──────────────────────────────────────────────────────────────
# Key-hex validation (used by the REST enroll-fc route)
# ──────────────────────────────────────────────────────────────

def parse_key_hex(key_hex: str) -> bytearray:
    """Parse a 64-char lowercase hex string into a 32-byte bytearray.

    Raises ValueError on any format error. Callers must treat the returned
    bytearray as sensitive and zeroize it after use.
    """
    if not isinstance(key_hex, str):
        raise ValueError("key_hex must be a string")
    if len(key_hex) != 64:
        raise ValueError(f"key_hex must be 64 hex chars, got {len(key_hex)}")
    try:
        raw = binascii.unhexlify(key_hex)
    except binascii.Error as exc:
        raise ValueError(f"key_hex is not valid hex: {exc}") from exc
    return bytearray(raw)

File: crates/ados-control/src/routes/signing.rs
//! MAVLink v2 signing capability + state read routes.
//!
//! The agent never holds a signing key. These read routes let the GCS detect
//! whether the connected FC supports MAVLink v2 signing and read the
//! observational signed-frame counters. Capability reads the vehicle-state
//! snapshot the MAVLink service publishes on `/run/ados/state.sock` (held by the
//! [`StateIpcClient`](crate::ipc)):
//!
//! - **`/api/mavlink/signing/capability`** runs the capability check over the FC
//!   connection flag and the autopilot id from the snapshot. ArduPilot keeps the
//!   signing key in its own persistent store, written with `SETUP_SIGNING`, and
//!   exposes no `SIGNING_*` parameter, so support is decided by the autopilot
//!   family, not by the param tree. Any other firmware reports unsupported with a
//!   specific reason enum.
//! - **`/api/mavlink/signing/counters`** reports signed-frame counters. The agent
//!   validates nothing (it holds no key); the counters would only confirm signed
//!   frames are transiting. No process on the node observes frames for signing
//!   today, so nothing is measured and the route says so: `observed: false` with
//!   every count `null`. A zero would read as "measured, and no signed frame was
//!   seen", which is a different and usually false claim.
//!
//! There is no signing-enforcement flag to read or set: ArduPilot has no
//! `SIGNING_*` parameter. Once its store holds a key it rejects unsigned traffic
//! on every link but channel 0 (normally USB), and it reports nothing about it.
//!
//! With no agent running (an empty snapshot), capability reports
//! `fc_not_connected` and counters report unmeasured, each a valid,
//! GCS-parseable body rather than a failure.

use axum::extract::State;
use axum::Json;
use serde_json::{json, Value};

use crate::state::AppState;

/// MAV_AUTOPILOT enum values the capability check distinguishes. ArduPilot is the
/// only firmware that gets the persistent signing store the GCS targets; PX4 and
/// the heartbeat-only / non-MAVLink case each get a specific reason. Mirrors the
/// `mavutil.mavlink.MAV_AUTOPILOT_*` constants the Python check reads.
const MAV_AUTOPILOT_ARDUPILOTMEGA: i64 = 3;
const MAV_AUTOPILOT_INVALID: i64 = 8;
const MAV_AUTOPILOT_PX4: i64 = 12;

/// `GET /api/mavlink/signing/capability` → whether the connected FC supports
/// MAVLink v2 signing.
///
/// Reads the FC connection flag and the autopilot id from the live state snapshot
/// and runs the capability check. Returns
/// `{supported, reason, firmware_name, firmware_version, signing_params_present}`.
/// `signing_params_present` is informational only (whether the cached param tree
/// carries any `SIGNING_*` name); it does not gate support. The `reason` enum:
/// `ok | fc_not_connected | firmware_not_supported |
/// firmware_px4_no_persistent_store | msp_protocol`. An absent snapshot reads as
/// disconnected → `fc_not_connected`. Guaranteed-200, never 500.
pub async fn capability(State(state): State<AppState>) -> Json<Value> {
    let snapshot = state.state.snapshot();
    let connected = fc_connected_from_snapshot(snapshot.as_ref());
    let autopilot = autopilot_from_snapshot(snapshot.as_ref());
    let params = crate::param_store::read_param_blob(&state.params_path);
    Json(detect_capability(
        connected,
        autopilot,
        signing_params_present(&params),
    ))
}

/// `GET /api/mavlink/signing/counters` → the observational signed-frame counters.
///
/// The agent holds no signing key and validates nothing. No process on the node
/// observes frames for signing, so nothing is measured: `{observed: false,
/// tx_signed_count: null, rx_signed_count: null, last_signed_rx_at: null}`.
pub async fn counters() -> Json<Value> {
    Json(json!({
        "observed": false,
        "tx_signed_count": Value::Null,
        "rx_signed_count": Value::Null,
        "last_signed_rx_at": Value::Null,
    }))
}

/// Read the FC connection flag out of a snapshot. Absent snapshot or absent flag
/// reads as disconnected, mirroring the Python `_fc_connected` (which reads the
/// state snapshot's `fc_connected`).
fn fc_connected_from_snapshot(snapshot: Option<&Value>) -> bool {
    snapshot
        .and_then(Value::as_object)
        .and_then(|m| m.get("fc_connected"))
        .and_then(Value::as_bool)
        .unwrap_or(false)
}

/// Read the MAV_AUTOPILOT id out of a snapshot's `autopilot` field. Absent or
/// non-numeric reads as `0`, mirroring the Python `_autopilot` (`int(ipc.get(
/// "autopilot", 0) or 0)`). A JSON `0`, a `null`, and an absent key all collapse
/// to `0` (the `or 0` in Python turns a falsy `0`/`None` into `0`).
fn autopilot_from_snapshot(snapshot: Option<&Value>) -> i64 {
    snapshot
        .and_then(Value::as_object)
        .and_then(|m| m.get("autopilot"))
        .and_then(Value::as_i64)
        .unwrap_or(0)
}

/// Whether any `SIGNING_*` param is present in the cached param map. Reported as
/// information only: ArduPilot's signing store is not a parameter, so a real
/// ArduPilot FC normally reports `false` here and is still supported.
fn signing_params_present(params: &serde_json::Map<String, Value>) -> bool {
    params.keys().any(|name| name.starts_with("SIGNING_"))
}

/// The capability check. The FC must be connected and the autopilot must be
/// ArduPilot, whose `SETUP_SIGNING` store is persistent. PX4, the MSP path and any
/// other autopilot report unsupported with a specific reason enum.
fn detect_capability(connected: bool, autopilot: i64, signing_params_present: bool) -> Value {
    if !connected {
        return json!({
            "supported": false,
            "reason": "fc_not_connected",
            "firmware_name": Value::Null,
            "firmware_version": Value::Null,
            "signing_params_present": false,
        });
    }

    let firmware_name = autopilot_name(autopilot);

    if autopilot != MAV_AUTOPILOT_ARDUPILOTMEGA {
        // An autopilot that isn't ArduPilot: PX4 has no persistent signing store,
        // the heartbeat-only / non-MAVLink case is the MSP path, anything else is
        // simply unsupported. (autopilot == 0 here means no heartbeat yet, which
        // is already covered above by the fc_not_connected branch.)
        let reason = if autopilot == MAV_AUTOPILOT_PX4 {
            "firmware_px4_no_persistent_store"
        } else if autopilot == MAV_AUTOPILOT_INVALID {
            "msp_protocol"
        } else {
            "firmware_not_supported"
        };
        return json!({
            "supported": false,
            "reason": reason,
            "firmware_name": firmware_name,
            "firmware_version": Value::Null,
            "signing_params_present": signing_params_present,
        });
    }

    json!({
        "supported": true,
        "reason": "ok",
        "firmware_name": firmware_name,
        // The GCS populates firmware_version when it reads AUTOPILOT_VERSION.
        "firmware_version": Value::Null,
        "signing_params_present": signing_params_present,
    })
}

/// The display name for a MAV_AUTOPILOT id, mirroring the Python `_autopilot_name`.
/// `0` → `null` (no heartbeat received yet); ArduPilot / PX4 / the heartbeat-only
/// case get their names; any other id gets the `Autopilot-<id>` form.
fn autopilot_name(autopilot: i64) -> Value {
    match autopilot {
        0 => Value::Null,
        MAV_AUTOPILOT_ARDUPILOTMEGA => json!("ArduPilot"),
        MAV_AUTOPILOT_PX4 => json!("PX4"),
        MAV_AUTOPILOT_INVALID => json!("Unknown (non-MAVLink or heartbeat-only)"),
        other => json!(format!("Autopilot-{other}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    // ── capability: the golden parity fixtures ───────────────────────────────

    #[test]
    fn capability_disconnected_is_the_fc_not_connected_default() {
        // The exact Python JSON for the disconnected case (no snapshot).
        let got = detect_capability(false, 0, false);
        assert_eq!(
            got,
            json!({
                "supported": false,
                "reason": "fc_not_connected",
                "firmware_name": Value::Null,
                "firmware_version": Value::Null,
                "signing_params_present": false,
            })
        );
    }

    #[test]
    fn capability_ardupilot_is_supported_without_signing_params() {
        // A real ArduPilot FC carries no SIGNING_* param: its key store is written
        // with SETUP_SIGNING. Support must not hinge on the param tree.
        let got = detect_capability(true, MAV_AUTOPILOT_ARDUPILOTMEGA, false);
        assert_eq!(
            got,
            json!({
                "supported": true,
                "reason": "ok",
                "firmware_name": "ArduPilot",
                "firmware_version": Value::Null,
                "signing_params_present": false,
            })
        );
    }

    #[test]
    fn capability_ardupilot_reports_signing_params_as_information() {
        let got = detect_capability(true, MAV_AUTOPILOT_ARDUPILOTMEGA, true);
        assert_eq!(got["supported"], json!(true));
        assert_eq!(got["signing_params_present"], json!(true));
    }

    #[test]
    fn capability_px4_has_no_persistent_store() {
        let got = detect_capability(true, MAV_AUTOPILOT_PX4, true);
        assert_eq!(
            got,
            json!({
                "supported": false,
                "reason": "firmware_px4_no_persistent_store",
                "firmware_name": "PX4",
                "firmware_version": Value::Null,
                "signing_params_present": true,
            })
        );
    }

    #[test]
    fn capability_invalid_autopilot_is_the_msp_path() {
        let got = detect_capability(true, MAV_AUTOPILOT_INVALID, false);
        assert_eq!(
            got,
            json!({
                "supported": false,
                "reason": "msp_protocol",
                "firmware_name": "Unknown (non-MAVLink or heartbeat-only)",
                "firmware_version": Value::Null,
                "signing_params_present": false,
            })
        );
    }

    #[test]
    fn capability_other_autopilot_is_unsupported_with_a_generic_name() {
        let got = detect_capability(true, 99, false);
        assert_eq!(
            got,
            json!({
                "supported": false,
                "reason": "firmware_not_supported",
                "firmware_name": "Autopilot-99",
                "firmware_version": Value::Null,
                "signing_params_present": false,
            })
        );
    }

    // ── snapshot field extraction ────────────────────────────────────────────

    #[test]
    fn fc_connected_reads_the_snapshot_flag() {
        let snap = json!({ "fc_connected": true });
        assert!(fc_connected_from_snapshot(Some(&snap)));
        let snap = json!({ "fc_connected": false });
        assert!(!fc_connected_from_snapshot(Some(&snap)));
        // Absent snapshot / absent flag both read as disconnected.
        assert!(!fc_connected_from_snapshot(None));
        assert!(!fc_connected_from_snapshot(Some(&json!({}))));
    }

    #[test]
    fn autopilot_reads_the_snapshot_id_and_defaults_to_zero() {
        let snap = json!({ "autopilot": 3 });
        assert_eq!(autopilot_from_snapshot(Some(&snap)), 3);
        // Absent snapshot / absent key / a null value all default to 0.
        assert_eq!(autopilot_from_snapshot(None), 0);
        assert_eq!(autopilot_from_snapshot(Some(&json!({}))), 0);
        assert_eq!(
            autopilot_from_snapshot(Some(&json!({"autopilot": null}))),
            0
        );
    }

    /// A `{name: value}` param map, the shape `crate::param_store` projects.
    fn param_map(body: Value) -> serde_json::Map<String, Value> {
        body.as_object().cloned().unwrap()
    }

    #[test]
    fn signing_params_present_scans_the_param_map() {
        // A SIGNING_* param present.
        assert!(signing_params_present(&param_map(
            json!({ "SIGNING_REQUIRE": 1.0, "FOO": 2.0 })
        )));
        // No SIGNING_* param.
        assert!(!signing_params_present(&param_map(json!({ "FOO": 2.0 }))));
        // An empty map — no cache file, or an FC that has answered nothing yet.
        assert!(!signing_params_present(&serde_json::Map::new()));
        // A name that merely CONTAINS the prefix is not a signing param.
        assert!(!signing_params_present(&param_map(
            json!({ "FOO_SIGNING_X": 1.0 })
        )));
    }

    // ── counters: the golden parity fixture ──────────────────────────────────

    #[tokio::test]
    async fn counters_report_unmeasured_rather_than_zero() {
        // No frame observer exists, so a count of 0 would be a fabricated
        // measurement. The body must say nothing was observed.
        let Json(body) = counters().await;
        assert_eq!(
            body,
            json!({
                "observed": false,
                "tx_signed_count": Value::Null,
                "rx_signed_count": Value::Null,
                "last_signed_rx_at": Value::Null,
            })
        );
    }
}
